User report: unable to use custom public key

From slack:

> I was not able to select a custom publickey for the cluster for some reason
> it would only let me use default ~/.ssh/id_rsa.pub
> when I tried another keys it crashed

Maybe a non-RSA key?

More details (thanks @felipejfc):

```
0705 15:30:49.555514   90244 executor.go:70] Tasks: 0 done / 47 total; 22 can run
I0705 15:30:50.515797   90244 vfs_castore.go:314] Issuing new certificate: "kubelet"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x125cff]

goroutine 81 [running]:
panic(0x11a9400, 0xc8200141b0)
    /usr/local/Cellar/go/1.6.2/libexec/src/runtime/panic.go:481 +0x3e6
k8s.io/kops/upup/pkg/fi.(*ResourceHolder).Open(0x0, 0x0, 0x0, 0x0, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/resources.go:165 +0x7f
k8s.io/kops/upup/pkg/fi.CopyResource(0x26047f8, 0xc8207fef50, 0x3a01b10, 0x0, 0x0, 0x0, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/resources.go:67 +0x57
k8s.io/kops/upup/pkg/fi.ResourceAsString(0x3a01b10, 0x0, 0x0, 0x0, 0x0, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/resources.go:85 +0xab
k8s.io/kops/upup/pkg/fi.DebugPrint(0x12f2ca0, 0x0, 0x0, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/values.go:44 +0xdc
k8s.io/kops/upup/pkg/fi.BuildChanges(0x13f3f80, 0xc820377320, 0x13f3f80, 0xc8203de640, 0x13f3f80, 0xc820377380, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/changes.go:64 +0x5ba
k8s.io/kops/upup/pkg/fi.DefaultDeltaRunMethod(0x22c2a98, 0xc8203de640, 0xc820548120, 0x0, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/default_methods.go:32 +0x5e2
k8s.io/kops/upup/pkg/fi/cloudup/awstasks.(*SSHKey).Run(0xc8203de640, 0xc820548120, 0x0, 0x0)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/cloudup/awstasks/sshkey.go:162 +0x35b
k8s.io/kops/upup/pkg/fi.(*executor).forkJoin.func1(0xc820510b00, 0x16, 0x16, 0xc8203ee5d0, 0xc820080048, 0xc820454a00, 0xa)
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/executor.go:141 +0x26c
created by k8s.io/kops/upup/pkg/fi.(*executor).forkJoin
    /Users/felipecavalcanti/dev/go/src/k8s.io/kops/upup/pkg/fi/executor.go:142 +0x160
```

> both keys are 2048bit RSA  keys
> one works and the other do not

@justinsb I figured it out, the problem was, I first created a cluster with key A, then I deleted it and then I tried to create a cluster with the same name using custom key B, and then it failed with this error:

**I0705 15:52:16.181515   93374 sshkey.go:77] Computed SSH key fingerprint mismatch: "xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx:xx" "yy:yy:y:yy:yy:yy:yy:yy:yy:yy:yy:yy:yy:yy:yy:yy"**
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x125cff]

This error then occurs if one key already exists on AWS and then you try to create a cluster with the same name with another key...

--- a/upup/pkg/fi/resources.go
+++ b/upup/pkg/fi/resources.go
@@ -162,6 +162,9 @@ type ResourceHolder struct {
 var _ Resource = &ResourceHolder{}
 
 func (o *ResourceHolder) Open() (io.ReadSeeker, error) {
+	if o.Resource == nil {
+		return nil, fmt.Errorf("ResourceHolder %q is not bound", o.Name)
+	}
 	return o.Resource.Open()
 }
 

--- a/upup/pkg/fi/values.go
+++ b/upup/pkg/fi/values.go
@@ -40,6 +40,11 @@ func DebugPrint(o interface{}) string {
 	if o == nil {
 		return "<nil>"
 	}
+	if rh, ok := o.(*ResourceHolder); ok {
+		if rh.Resource == nil {
+			return fmt.Sprintf("unknown resource %q", rh.Name)
+		}
+	}
 	if resource, ok := o.(Resource); ok {
 		s, err := ResourceAsString(resource)
 		if err != nil {

